example/kalman_filter/kalman_filter-example_2.cpp

```diff
@@ -46,15 +48,17 @@ void print(auto iteration, Quantity auto measured, const kalman::State auto& cur
 
 int main()
 {
-  using namespace units::isq;
-  using namespace units::isq::si::references;
-  using state = kalman::state<si::length<si::metre>, si::speed<si::metre_per_second>>;
+  using namespace units::si::unit_symbols;
+  using state = kalman::state<quantity<isq::position_vector[m]>, quantity<isq::velocity[m / s]>>;
 
-  const auto interval = 5 * s;
-  const state initial = {30 * km, 40 * (m / s)};
-  const std::array measurements = {30110 * m, 30265 * m, 30740 * m, 30750 * m, 31135 * m,
-                                   31015 * m, 31180 * m, 31610 * m, 31960 * m, 31865 * m};
-  std::array gain = {dimensionless<one>(0.2), dimensionless<one>(0.1)};
+  const auto interval = 5 * isq::period_duration[s];
+  const state initial = {30 * isq::position_vector[km], 40 * isq::velocity[m / s]};
```

Comment:
```diff
using state = kalman::state<quantity<isq::position_vector[m]>, quantity<isq::velocity[m / s]>>;
-const state initial = {30 * km, 40 * (m / s)};
+const state initial = {30 * isq::position_vector[km], 40 * isq::velocity[m / s]};
```
seems like an unfortunate deviation from the general idea of ISO 80000-1, 7.1.4 “Expression for quantities” and 7.2 “Names and symbols for units”'s 7.2.1 “General”. Since `state` is already strongly typed, could the previous iterator be made to work? I think it's preferable if C++ code looked like the quantity expression. Strong typing could be added without affecting it.

Reply:
> Strong typing could be added without affecting it.

More like the quantity.

> [...] It is therefore important not to use the unit alone to specify the quantity. This applies not only to
technical texts, but also, for example, to measuring instruments (i.e. the instrument read-out
needs to indicate both the unit and the quantity measured).
-- Extract from https://www.bipm.org/en/publications/si-brochure

The intention seems to be for "30 km" to not be decorated with anything other than the number and the unit. In C++, we use `*` out of necessity. But this shouldn't be the place to specify the quantity.

Reply:
As you know, a number and a unit are not enough to define a quantity. Even adding a dimension was found to be deficient. Also, as the quantity cannot be implicitly constructible from the representation type, I cannot just leave numbers to initialize the struct.

I understand that you would like to allow `30 * km` here, but what should it do? We cannot create a quantity from that. We do not know the quantity type/kind based on this, and I am not sure if that is not a big issue here. A unit is probably overhead and duplication here unless we want to allow initialization with a different unit than a unit provided for a quantity. But for that, we will have to introduce yet another type/abstraction that would handle such a scenario and make a `quantity` implicitly constructible from it. Is it worth it?

Alternatively, we could consider allowing implicit initialization of a quantity from a value, but is it safe in all cases (i.e. passing a number as a function argument)?

Do you think that it is a good idea to introduce one more

Reply:
Please let me know what are your ideas here and we can discuss and possibly implement them.

Reply:
> I understand that you would like to allow `30 * km` here, but what should it do?

In the context of the commented code, `state` already has concrete quantities, so `30 * km` would initialize it one of them.

> A unit is probably overhead and duplication here unless we want to allow initialization with a different unit than a unit provided for a quantity. But for that, we will have to introduce yet another type/abstraction that would handle such a scenario and make a `quantity` implicitly constructible from it. Is it worth it?

I don't know. That question could be added to the standardization effort. What I know is that the symbols lose some amount of their worth if they have to be surrounded by long quantity names.

Reply:
> In the context of the commented code, state already has concrete quantities, so 30 * km would initialize it one of them.

The same as just `30`, which would work if a quantity was implicitly constructible from a value. `km` could be considered an overhead here.

Reply:
Ohh, and there is one more point here. Right now we do the following to define a scaled unit:

```cpp
inline constexpr struct minute : named_unit<"min", mag<60> * second> {} minute;
```

and I hope that if expression aliases would become a part of the C++ language we would write:

```cpp
inline constexpr struct minute : named_unit<"min", 60 * second> {} minute;
```

so a `60 * s` is just yet another scaled unit and not a quantity value. Is it a bad idea? In the long run, I would like to think about how to simplify or even remove the use of `magnitude` complex engine in units definition. If you have any ideas or experience here, you are welcome to share your thoughts or prototype something as I know you worked on this already a bit.

Reply:
I was thinking of permitting `x * km`, where `x` is a function parameter, to initialize any quantity of dimension L (besides the requirements on the representation and the unit).

Reply:
Regarding function parameter, I am not the biggest fan of this syntax in general. Consider:

```cpp
void foo(double a, double b)
{
  isq::distance[km] l = a * m;
  isq::duration[h] d = b * s;
  units::isq::speed<km / h> speed = l / d;
}
```

This looks really cryptic to me and might be a source of confusion. I would much prefer people to type:

```cpp
void foo(double a, double b)
{
  auto l = a * isq::distance[m];
  auto d = b * isq::duration[s];
  units::isq::speed<km / h> speed = l / d;
}
```

So I do not think that there is a problem here.

The problem that we discuss is a different one. It is an inability to efficiently initialize aggregates of quantities:

```cpp
quantity<isq::length[m]> vec[3] = { 1 * isq::length[m], 2 * isq::length[m], 3 * isq::length[m] };
std::tuple<quantity<isq::length[m]>, quantity<isq::time[s]>> t = { 1 * isq::length[m], 1 * isq::time[s] };
```

If we would make the constructor implicit than the above could be written as:

```cpp
quantity<isq::length[m]> vec2[3] = { 1, 2, 3 };
std::tuple<quantity<isq::length[m]>, quantity<isq::time[s]>> t2 = { 1, 1 };
```

But would also allow the following:

```cpp
void foo(quantity<isq::length[m]>);

foo(3);
```

which I am not so sure if it is a great idea, although I am ready to discuss and consider it.

Reply:
I also believe that the creation of quantities from raw number arguments is quite rare and happens only on the boundaries of strongly typed engines. On the other hand, the initialization of some quantities from constant values will happen quite often and sometimes a quantity will be packed inside of the aggregate.

Reply:
> The problem that we discuss is a different one. It is an inability to efficiently initialize aggregates of quantities:

Not really. I just happened to notice thanks to this example. For me, it's all about `3 * isq::length[km]` being too far from "3 km".

Reply:
I think that is what we have to deal with here. How otherwise you will specify that you want to create a quantity of `length`, `distance`, `width`, or `height`? `3 * m` is not enough to define a quantity.

Reply:
If in a specific project we use only one kind of quantity and we use it often we can always define a helper like:

```cpp
constexpr auto kmph = units::isq::speed[km / h];
Quantity auto speed = 60 * kmph;
```

Reply:
Specify the quantity somewhere else. E.g.
- The radius of a circle is 3 m.
  `quantity<radius, centimetre> radius_of_circle = 3 * m;`.
- 𝘤 = 299 792 458 m s⁻¹.
  `unit_constant<speed, ???> speed_of_light = 299 792 458 * (m * pow<-1>(s));`.

Reply:
Alternatively, specify that the intention of unit symbols is not for C++ expressions to resemble Math and Physics, like `std::chrono::literals` resemble expressions of time.

Reply:
The first case is interesting indeed, but I am not sure if enough to introduce a new abstraction to the engine and also limit possible simplification of unit definitions with expression aliases.  A good point here is that we can always add it later.

Regarding the second case, `unit_constant` is to produce a unit that has the value embedded in the type which is impossible with the syntax provided above. Doing that proves to be really powerful (for examples see _total_energy_ or `fill_level()` from _box_example_).

Reply:
Oh, right. That's how it works. I didn't mean to suggest changing `unit_constant`. Just that any expression 𝘌 like `1 * km / (2 * s)` should be usable to implicitly initialize a quantity of dimension dim 𝘌.

Reply:
Mostly. `1 * Hz` should hopefully error to initialize anything other than a quantity of frequency.

Reply:
> Alternatively, specify that the intention of unit symbols is not for C++ expressions to resemble Math and Physics, like std::chrono::literals resemble expressions of time.

For me, `unit_symbols` (or previously `reference`s) was never about the exact resemblance of Physics but about easy and efficient construction of strong types that represent a specific quantity. As we agreed, a quantity is much more than just a value and a unit so it cannot be efficiently defined with this syntax. Another point, as you stated, is when we have a quantity definition somewhere already and we want to initialize it using a different unit. I agree that this case looks interesting and could be considered as a future extension. We can open an Issue for that.

However, this still does not address the original issue. One could still argue that, even with the feature proposed by you, the `q2` below is too verbose:

```cpp
quantity<isq::speed[km / h]> q1{60};
quantity<isq::speed[km / h]> q2[3] = { 10 * (km / h), 20 * km / h), 30 * (km / h) };
```

This is why I opened #410 for that.

Reply:
> Mostly. 1 * Hz should hopefully error to initialize anything other than a quantity of frequency.

This is yet another issue. I do not see any way to restrict that. For now, you can use `[Hz]` for any quantity kind that has a dimension of `1 / dim_time`. If you have any ideas on how to improve that without heavily complicating the unit system definition, please let me know.

Reply:
> For me, `unit_symbols` (or previously `reference`s) was never about the exact resemblance of Physics but about easy and efficient construction of strong types that represent a specific quantity.

Then it's doing its job pretty well. We just have to make sure that's appropriately documented.

Reply:
> If you have any ideas on how to improve that without heavily complicating the unit system definition, please let me know.

How about a wrapper around the `unit_symbol` that additionally takes the quantity the unit is restricted to. Then the definition of `Hz` would only need to change `unit_symbol` to the name of the new abstraction, and additionally specify `frequency` as a template parameter.

Reply:
> By default, you can only specify a quantity with long unit names that do not make a resemblance to Physics, right?

Of course. But the existence of `reference`s made it possible. And I was sad to see it go.

Reply:
Do you mean something like that:

```cpp
inline constexpr struct hertz : named_unit<"Hz", 1 / second, only_for<isq::frequency>> {} hertz;
```

That might be a good idea. But it should match not only frequency but also all kinds of it. Moreover, it should not match `1 / isq::time`. We can try to implement that...

Reply:
> But the existence of references made it possible. And I was sad to see it go.

Yes, that was an unfortunate effect of removing the downcasting facility, which allowed us to model many quantity kinds with just a `quantity` class template. But fortunately, as I stated above, any user can easily provide a custom helper for a reference instance if the kind is clearly known from the identifier in a specific project.

Reply:
BTW, I already did it here https://github.com/mpusz/units/blob/v2_discussion/docs/v2_design.md#quantity-reference.

Reply:
How about this syntax? The call stack size would be the same, so overhead shouldn't be an argument against it.

N | Current | Proposed
---|---|---
1 | `30 * isq::speed[m / s]` | `isq::speed(30, m / s)`
2 | `60 * isq::length[m] / (2 * isq::time[s])` | `isq::length(60, m) / isq::time(2, s)`
3 | `60 * isq::length[m] / (2 * isq::time[s] * (x * isq::time[s]))` | `isq::length(60, m) / (isq::time(2, s) * isq::time(x, s))`
4 | `60 * isq::length[m] / ((2 * isq::time[s]) * (x * isq::time[s]))` | `isq::length(60, m) / (isq::time(2, s) * isq::time(x, s))`
5 | `y * isq::mass[kg] * (z * isq::length[m]) * (w * isq::time[s])` | `isq::mass(y, kg) * isq::length(z, m) * isq::time(w, s)`

- `isq::speed(30, km / h)` has the number and unit side-by-side, and the quantity at the left, a very common order.
- `isq::length(60, km) / isq::time(2, h)` requires less parentheses for associativity. So do 3 and 4.
- `[]` for indexing is replaced by `()` for calling, and there's no omitting them unlike the `()` for associativity in the first quantity of an expression, i.e.
  `x₀ * q₀[u₀] * (x₁ * q₁[u₁])` and
  `(x₀ * q₀[u₀]) * (x₁ * q₁[u₁])` both work, but in the proposed syntax it's always
  `q₀(x₀, u₀) * q₁(x₁, u₁)`.
- 5 is to get a feel on the amount of `*`s (not) required.

Reply:
I must say it is an interesting idea. However, how would I redefine that?

```cpp
quantity<isq::length[m]> q1{42};
quantity_of<isq::length[m]> auto q2  = q1;
auto q3 = quantity_cast<isq::distance[km]>(q1);
```

Reply:
I suppose you could just omit the first argument, the number, i.e. `isq::length(m)`.

Reply:
Assuming that `isq::length(m)` would create a `reference` how can we create a quantity from a `reference` and a `value`? Should we leave the multiply syntax or provide a call operator like:

```cpp
constexpr auto kmph = isq::speed(km / h);
Quantity auto q = kmph(60);
```

Reply:
Or maybe we should get rid of references and end up with:

```cpp
quantity<isq::length, m> q1{42};
quantity_of<isq::length, m> auto q2  = q1;
auto q3 = quantity_cast<isq::distance, km>(q1);
```

?

Reply:
But then we will always need to specify a quantity type like `distance`, `height`, etc, and will not be able to do https://github.com/mpusz/units/pull/391#discussion_r1054698532.

Reply:
Also, maybe because of the amount of time I spent with the new syntax already, I find the first line easier to parse and prove correct than the second one. For me at least, it seems easier to find and analyze in isolation a number and a quantity in the first line.

```cpp
static_assert(120 * isq::distance[km] / (2 * isq::duration[h]) == 60 * isq::speed[km / h]);
static_assert(isq::distance(120, km) / isq::duration(2, h) == isq::speed(60, km / h));
```

Reply:
> But then we will always need to specify a quantity type like `distance`, `height`, etc, and will not be able to do [#391 (comment)](https://github.com/mpusz/units/pull/391#discussion_r1054698532).

Then, omitting the number can just behave as `𝘲[𝘶]` currently does. So `𝘲(𝘶)` and `𝘲[𝘶]` are equivalent, and no further changes are needed.

Reply:
Yes, but what to do with the `reference` being a result of that operation? For now, it was mandatory part to create and define quantities. With the proposed change, it gets less important. So the question is if maybe we should remove it.

If we remove it, we will not be able to do a few things:
1. `quantity` will not be specified in terms of a number and a `reference` but `quantity_spec` and `unit`.
2. `quantity_cast` and `quantity_of` would require two parameters in this case which complicates the design.
3. We will not be able to do things like `interconvertible(isq::length[km] / isq::time[h], isq::speed[m/s])`.

If we leave it, we have to find out how to construct a quantity from it:
4. Should we leave a multiply syntax or provide a call operator? The first case is probably a bad idea as this syntax has too many issues to standardize it only to be used in just a few places as most users will use another call operator to create a quantity.

Also, it bothers me a bit that it might be hard to teach and learn if we will have a lot of those call operators:
- `Quantity auto QuantitySpec::operator()(Representation auto, Unit auto);`
- `Reference auto QuantitySpec::operator()(Unit auto);`
- `Quantity auto Reference::operator()(Representation auto);`

But I see your point and am open to making such a change if we resolve all the issues and agree that the change is for the better.

Reply:
Of course, we can provide both syntaxes as well as alternative experimental features, as it is the case of the mainline, where we have four of those already (https://mpusz.github.io/units/framework/quantities.html#quantity-construction-helpers). But I am not sure if this is the right way to go...

Reply:
It's better to have a single syntax. I'm assuming you're referring to `q[u]` vs `q(u)`, and not whether `q(u)` should be a thing.

Reply:
I am referring to having both of the below for a while:

```cpp
static_assert(120 * isq::distance[km] / (2 * isq::duration[h]) == 60 * isq::speed[km / h]);
static_assert(isq::distance(120, km) / isq::duration(2, h) == isq::speed(60, km / h));
```

Reply:
I did some prototyping and it seems that the best way is to have:
- `Quantity auto Reference::operator()(Representation auto);`
- `Reference auto QuantitySpec::operator[](Unit auto);`
- `Quantity auto QuantitySpec::operator()(Representation auto v, Unit auto u) { return Self[u](v); }`

To say it otherwise, I would prefer to leave `quantity_ref[unit]` to create a `reference`, as I think that `isq:::length[m](42)` is more readable than `isq::length(m)(42)` and it also looks so nice in all other cases where we need references (i.e. quantity<isq::length[m]>). 😉

This means that the biggest change is to replace multiply syntax for references with a call operator.

Reply:
Actually, while refactoring some examples, I realized that it could get even messier. The following syntax is also valid `isq::length[m](42)[mm]`, which creates a quantity of `isq::length[m]` and a number `42`, and converts it to a quantity in millimeters.

Reply:
With https://github.com/mpusz/units/pull/391#discussion_r1054723916 in mind, that seems like an improvement.

Reply:
I just pushed new syntax changes as well as refactoring of some examples. Please let me know if that is what you had in mind and if you find this syntax superior.

Reply:
I reviewed the change at https://github.com/mpusz/units/pull/391#discussion_r1055674150. Previously, you mentioned that an advantage of the `x * q[u]` syntax is that the same `q[u]` is consistently used in `quantity<q[u]>` and `quantity_cast<q[u]>`. For my suggested `q(2, u)` syntax, I suggested `q(u)` as a replacement, which isn't as consistent, and to make `q(2, u)` and `q(u)(x)` equivalent. Since the refactor allows both syntaxes, there were no uses of `q(u)` to review. Although there were `r(x)`, where `r` is a variable with value `q(u)` (actually initialized with `q[u]`).

Reply:
Again with https://github.com/mpusz/units/pull/391#discussion_r1054723916 in mind, `isq:::length[m](42)` seems like a good choice. Specially if one won't use the `unit_symbols` namespace, i.e. `isq:::length[meter](42)`. In this last case, I currently consider it superior to `isq::length(42, metre)`. If `isq::length[m](42)[mm]` being valid was such a big deal, it could be removed. But it seems like the kind of thing that would only turn out in unit tests, where actual use cases would subscript a quantity variable.

Reply:
I also tend to prefer `isq:::length[metre](42)` over `isq::length(42, metre)`. Regarding the shorter ones `isq:::length[m](42)` and `isq::length(42, m)`, I am undecided as I have gotten used to the `isq:::length[m]` syntax a lot through the last 3 months working on V2, and this coma in the other expression still looks strange to me and may be prone to ordering issues.

Please let me know how you prefer to progress here. For sure, we need `quantity_spec[]` and `reference()` as basic blocks. We also need to remove multiply syntax from the `reference`. The question is if we should provide a two-argument call operator in `quantity_spec`. If needed, it can always be easily added later...

Reply:
> Please let me know how you prefer to progress here. For sure, we need `quantity_spec[]` and `reference()` as basic blocks. We also need to remove multiply syntax from the `reference`.

Let's go with this for now.

Reply:
This also helps in consistency "put a unit in a bracket after the quantity type/kind".